File: sourcecode/Convolve.js
const {performance} = require('perf_hooks');

function f(n) {
    return 1/n;
}

function g(n) {
    return Math.sin(n);
}

function convolve(f,g,k) {
    let sum = 0.0;
    for (let i = 1; i <= k; i++) {
        //if (i < 0) break;
        if (k % i === 0) {
            sum += f(i) * g(k/i);
        }
    }
    return sum;
}

function convolve2(f,g,k) {
    let sum = 0.0;
    for (let i = 1; i*i <= k; i++) {
        //if (i < 0) break;
        if (k % i === 0) {
            sum += f(i) * g(k/i);
            if (i*i !== k) sum += f(k/i) * g(i);
        }
    }
    return sum;
}

function convolve3(f,g,k) {
    return Array
        .from(Array(k).keys())
        .map((i) => i+1)
        .filter((item) => k % item === 0)
        .reduce((sum, item) => sum + f(item) * g(k/item), 0)
}

function test1(k) {
    let start = performance.now(); 
    let a = convolve(f,g,k);
    let end = performance.now(); 
    return (end - start).toFixed(6).toString();
}

function test2(k) {
    let start = performance.now(); 
    let a = convolve2(f,g,k);
    let end = performance.now(); 
    return (end - start).toFixed(6).toString();
}

function test3(k) {
    if (k < 100000000) {
        let start = performance.now(); 
        let a = convolve3(f,g,k);
        let end = performance.now(); 
        return (end - start).toFixed(6).toString();
    } else {
        return "OffTheScale"
    }
}

var myArgs = process.argv.slice(2);

//console.log(convolve(f,g,2137))

process.stdout.write(
    test1(parseInt(myArgs[0]))
    +"\t"+test2(parseInt(myArgs[0]))
    +"\t"+test3(parseInt(myArgs[0]))
);

File: sourcecode/Convolve.py
import math
import time
import sys

def fix(a):
    return "{:.6f}".format(a)

def f(n):
    return 1.0/n

def g(n):
    return math.sin(n)

def convolve(f,g,n):
    sum = 0.0
    for d in range(1, n+1):
        if (n % d == 0):
            sum += f(d)*g(n//d)
    return sum

def convolve2(f,g,n):
    sum = 0.0
    for d in range(1, math.trunc(math.sqrt(n+1))):
        if (n % d == 0):
            sum += f(d)*g(n//d)
            if (d * d != n):
                 sum += f(n//d)*g(d)
    return sum


def convolve3(f,g,n):
    return sum(f(d)*g(n//d) for d in range(1,n+1) if 1>n%d)

def test(k):
    start = time.time_ns()
    a = convolve(f,g,k)
    end = time.time_ns()
    return fix((end-start)/1000000.0)

def test2(k):
    start = time.time_ns()
    a = convolve2(f,g,k)
    end = time.time_ns()
    return fix((end-start)/1000000.0)

def test3(k):
    start = time.time_ns()
    a = convolve3(f,g,k)
    end = time.time_ns()
    return fix((end-start)/1000000.0)

if __name__ == '__main__':
    k = int(sys.argv[1])
    print(test(k), test2(k), test3(k), end='', sep='\t')


